aggregator/src/datastore.rs

```diff
@@ -2149,6 +2149,92 @@ WITH updated as (
+        HAVING bool_and(aggregation_jobs.state != 'IN_PROGRESS')
+        -- Honor maximum_acquire_count *after* winnowing down to runnable collect jobs
+        LIMIT $3
+    )
+    RETURNING tasks.task_id, tasks.query_type, tasks.vdaf, collect_jobs.collect_job_id,
+              collect_jobs.id, collect_jobs.lease_token, collect_jobs.lease_attempts
+)
+SELECT task_id, query_type, vdaf, collect_job_id, lease_token, lease_attempts FROM updated
+-- TODO (#174): revisit collect job queueing behavior implied by this ORDER BY
+ORDER BY id DESC
+"#,
+            )
+            .await?;
+        self.tx
+            .query(
+                &stmt,
+                &[
+                    /* lease_expiry */ &lease_expiry_time.as_naive_date_time()?,
+                    /* now */ &now.as_naive_date_time()?,
+                    /* limit */ &maximum_acquire_count,
+                ],
+            )
+            .await?
+            .into_iter()
+            .map(|row| {
+                let task_id = TaskId::get_decoded(row.get("task_id"))?;
+                let collect_job_id = row.get("collect_job_id");
+                let query_type = row.try_get::<_, Json<task::QueryType>>("query_type")?.0;
+                let vdaf = row.try_get::<_, Json<VdafInstance>>("vdaf")?.0;
+                let lease_token_bytes: [u8; LeaseToken::LEN] = row
+                    .get::<_, Vec<u8>>("lease_token")
+                    .try_into()
+                    .map_err(|err| Error::DbState(format!("lease_token invalid: {:?}", err)))?;
+                let lease_token = LeaseToken::from(lease_token_bytes);
+                let lease_attempts = row.get_bigint_and_convert("lease_attempts")?;
+                Ok(Lease::new(
+                    AcquiredCollectJob::new(task_id, collect_job_id, query_type, vdaf),
+                    lease_expiry_time,
+                    lease_token,
+                    lease_attempts,
+                ))
+            })
+            .collect()
+    }
+
+    /// acquire_incomplete_fixed_size_collect_jobs retrieves & acquires the IDs of unclaimed
+    /// incomplete collect jobs. At most `maximum_acquire_count` jobs are acquired. The job is
+    /// acquired with a "lease" that will time out; the desired duration of the lease is a
+    /// parameter, and the lease expiration time is returned. Applies only to fixed-size tasks.
+    pub async fn acquire_incomplete_fixed_size_collect_jobs(
+        &self,
+        lease_duration: &Duration,
+        maximum_acquire_count: usize,
+    ) -> Result<Vec<Lease<AcquiredCollectJob>>, Error> {
+        let now = self.clock.now();
+        let lease_expiry_time = now.add(lease_duration)?;
+        let maximum_acquire_count: i64 = maximum_acquire_count.try_into()?;
+
+        let stmt = self
+            .tx
+            .prepare_cached(
+                r#"
+WITH updated as (
+    UPDATE collect_jobs SET lease_expiry = $1, lease_token = gen_random_bytes(16), lease_attempts = lease_attempts + 1
```

Comment:
I'd like to see logic like this re-used between the fixed size and time interval queries, since the main difference between the two seems to be the `INNER JOIN` against `aggregation_jobs`. However, I'm not sure what's a good way to reuse code between prepared statements so I'm OK going ahead with this.

Reply:
Yeah, I played with this for a bit. We could do literal string concatenation, but that doesn't feel particularly maintainable. More than that would require an ORM (or the moral equivalent), and I don't think we want to go that way.